test/tool-option-parsing.py

```diff
@@ -104,200 +99,182 @@ def run_command_missing_arg(self, args):
-def test_how_to_type(xkbcli_how_to_type, args):
-    xkbcli_how_to_type.run_command_success(args)
-
-
-@pytest.mark.parametrize('rmlvos', rmlvos)
-def test_how_to_type_rmlvo(xkbcli_how_to_type, rmlvos):
-    args = rmlvos + ['0x1234']
-    xkbcli_how_to_type.run_command_success(args)
-
-
-@pytest.mark.parametrize('args', [['--verbose'],
-                                  ['-v'],
-                                  ['--verbose', '--load-exotic'],
-                                  ['--load-exotic'],
-                                  ['--ruleset=evdev'],
-                                  ['--ruleset=base'],
-                                  ])
-def test_list_rmlvo(xkbcli_list, args):
-    xkbcli_list.run_command_success(args)
-
-
-def test_list_rmlvo_includes(xkbcli_list):
-    args = ['/tmp/']
-    xkbcli_list.run_command_success(args)
-
-
-def test_list_rmlvo_includes_invalid(xkbcli_list):
-    args = ['/proc/version']
-    rc, stdout, stderr = xkbcli_list.run_command(args)
-    assert rc == 1
-    assert "Failed to append include path" in stderr
-
-
-def test_list_rmlvo_includes_no_defaults(xkbcli_list):
-    args = ['--skip-default-paths', '/tmp']
-    rc, stdout, stderr = xkbcli_list.run_command(args)
-    assert rc == 1
-    assert "Failed to parse XKB description" in stderr
-
-
-@pytest.mark.skipif(not os.path.exists('/dev/input/event0'), reason='event node required')
-@pytest.mark.skipif(not os.access('/dev/input/event0', os.R_OK), reason='insufficient permissions')
-@pytest.mark.parametrize('rmlvos', rmlvos)
-def test_interactive_evdev_rmlvo(xkbcli_interactive_evdev, rmlvos):
-    return
-    xkbcli_interactive_evdev.run_command_success(rmlvos)
-
-
-@pytest.mark.skipif(not os.path.exists('/dev/input/event0'),
-                    reason='event node required')
-@pytest.mark.skipif(not os.access('/dev/input/event0', os.R_OK),
-                    reason='insufficient permissions')
-@pytest.mark.parametrize('args', [['--report-state-changes'],
-                                  ['--enable-compose'],
-                                  ['--consumed-mode=xkb'],
-                                  ['--consumed-mode=gtk'],
-                                  ['--without-x11-offset'],
-                                  ])
-def test_interactive_evdev(xkbcli_interactive_evdev, args):
-    # Note: --enable-compose fails if $prefix doesn't have the compose tables
-    # installed
-    xkbcli_interactive_evdev.run_command_success(args)
-
-
-@pytest.mark.skipif(not os.getenv('DISPLAY'), reason='DISPLAY not set')
-def test_interactive_x11(xkbcli_interactive_x11):
-    # To be filled in if we handle something other than --help
-    pass
-
-
-@pytest.mark.skipif(not os.getenv('WAYLAND_DISPLAY'),
-                    reason='WAYLAND_DISPLAY not set')
-def test_interactive_wayland(xkbcli_interactive_wayland):
-    # To be filled in if we handle something other than --help
-    pass
+class TestXkbcli(unittest.TestCase):
```

Comment:
if we're using unittest we should use the one thing it does better than pytest: hierarchy. So one generic `XkbCliBase` class with the `help` and `invalid-option` test, then a subclass of that for every tool. That's what I did in libinput first, see [the switch to pytest commit](https://gitlab.freedesktop.org/libinput/libinput/-/commit/d0eb8d58a3e3c1f1f15e612c732a43486c6479e4) which gives you a nice before/after, you just need to invert this :)

That way it'll be easier to skip a test set too if the registry isn't compiled. Sorry about that bit.

Reply:
> if we're using unittest we should use the one thing it does better than pytest: hierarchy

You can use test classes in pytest as well :wink:

> That's what I did in libinput first, see the switch to pytest commit which gives you a nice before/after, you just need to invert this :)

Thanks, I'll check it out.

Reply:
I ended up not splitting into classes, but made some hack to make it skip more easily. I think it's OK.